jaraco/develop/git.py

```diff
@@ -109,29 +107,75 @@ def path(self):
 
 class Project(str):
     """
-    >>> p = Project.parse('foo-project [tag1] [tag2]')
+    >>> p = Project.parse('foo-project [tag1] [tag2] (zero defect, coherent software)')
     >>> p
     'foo-project'
     >>> p.tags
     ['tag1', 'tag2']
     >>> p.spec
     'foo-project [tag1] [tag2]'
+    >>> p.topics
+    ['zero defect', 'coherent software']
     """
 
     pattern = re.compile(r'(?P<name>\S+)\s*(?P<rest>.*)$')
     tags: list[str] = []
-
-    def __new__(self, value, **kwargs):
-        return super().__new__(self, value)
+    cache: dict[str, Project] = {}
+
+    def __new__(cls, value, **kwargs):
+        # Down-cast to a string early.
+        value = sys.intern(str(value))
+        try:
+            return cls.cache[value]
+        except KeyError:
+            new = super().__new__(cls, value)
+            cls.cache[new] = new
+            return new
```

Comment:
Why not:

```suggestion

    @functools.lru_cache
    def __new__(cls, value, **kwargs):
        return super().__new__(cls, value)
```

I'm not sure what is the purpose of `value = sys.intern(str(value))`. Are there situations where value isn't already a string? Are we not potentially optimizing too early? If this project is only ever dealing with a few hundred projects, I'd expect the value of caching and interning to be negligible.

Reply:
I (vaguely) tried to explain the purpose of `value = sys.intern(str(value))` in https://github.com/jaraco/jaraco.develop/pull/10#issuecomment-1925454025: the argument `value` might be a string subclass, so we cast it just in case. However, I do think `@functools.lru_cache` might be much better here, since hashes of `str` instances and `str` subclass instances are the same for the same initializing string.